xla/service/collective_quantizer.cc

```diff
@@ -50,14 +60,121 @@ auto ConvertToNarrowerType() {
   return m::Convert().WithPredicate(converts_to_narrower_type);
 }
 
+// Matches a type conversion to a type with a larger byte size than that of the
+// operand.
+auto ConvertToWiderType() {
+  auto converts_to_wider_type = [](const HloInstruction* instr) -> bool {
+    return ShapeUtil::ByteSizeOfPrimitiveType(instr->shape().element_type()) >
+           ShapeUtil::ByteSizeOfPrimitiveType(
+               instr->operand(0)->shape().element_type());
+  };
+  return m::Convert().WithPredicate(converts_to_wider_type);
+}
+
+bool IsSupportedCollective(HloInstruction* instr) {
+  return instr->operand_count() == 1 &&
+         (instr->opcode() == HloOpcode::kAllGather ||
+          instr->opcode() == HloOpcode::kAllToAll ||
+          instr->opcode() == HloOpcode::kCollectiveBroadcast ||
+          instr->opcode() == HloOpcode::kCollectivePermute);
+}
+
+// Sequentially applies the ops in unaries to the output of instr.
+HloInstruction* ApplyUnary(HloInstruction* instr,
+                           const std::vector<HloInstruction*>& unaries) {
+  for (HloInstruction* unary_op : unaries) {
+    instr = instr->AddInstruction(unary_op->CloneWithNewOperands(
+        ShapeUtil::MakeShapeWithDenseLayout(
+            instr->shape().element_type(), unary_op->shape().dimensions(),
+            unary_op->shape().layout().minor_to_major()),
+        {instr}));
+  }
+  return instr;
+}
+
+// Recursively collects unary, divide, or multiply operands of instr until a
+// conversion to a wider type is reached. Returns an empty vector when no
+// conversion is reached.
+std::vector<HloInstruction*> FindDequantizationSubgraphRecursive(
```

Comment:
Can you write this function in a similar style to the [corresponding function](https://github.com/openxla/xla/blob/676521539300ce2141c169b314d293bda920c277/xla/service/gpu/transforms/gemm_rewriter.cc#L200) in gemm_rewriter.cc? A few months ago, I refactored the function to make it more clear. I think you can just copy the new version with some minor changes, like using `std::vector<HloInstruction*>` instead of `std::vector<std::pair<HloInstruction *, int>>`.

In particular, the comment on the gemm_rewriter function is more explicit about what is done, and the order of HLO instructions doesn't need to be reversed, and the subgraph shouldn't be both passed-in and returned.